substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,609 @@
+                if (analysis) {
+                    map.putIfAbsent(invokeBci, InvocationResult.NO_ANALYSIS);
+                }
+            }
+        }
+    }
+
+    private final boolean analysis;
+    private final HostedProviders providers;
+    private static int numberForInline = 0;
+    private static final ConcurrentHashMap<AnalysisMethod, ConcurrentHashMap<Integer, InvocationResult>> dataInline = new ConcurrentHashMap<>();
+
+    public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders providers) {
+        this.analysis = analysis;
+        this.providers = providers;
+    }
+
+    public static int getNumberForInline() {
+        return numberForInline;
+    }
+
+    private static ConcurrentMap<Integer, InvocationResult> bciMap(ResolvedJavaMethod method) {
+        AnalysisMethod key;
+        if (method instanceof AnalysisMethod) {
+            key = (AnalysisMethod) method;
+        } else {
+            key = ((HostedMethod) method).getWrapped();
+        }
+
+        return dataInline.computeIfAbsent(key, unused -> new ConcurrentHashMap<>());
+    }
+
+    private static InvocationResult getResult(ResolvedJavaMethod method, int bci) {
+        return bciMap(method).get(bci);
+    }
+
+    public static InvocationResult putIfAbsent(ResolvedJavaMethod method, int bci, InvocationResult value) {
+        return bciMap(method).putIfAbsent(bci, value);
+    }
+
+    @Override
+    public InlineInfo shouldInlineInvoke(GraphBuilderContext b, ResolvedJavaMethod method, ValueNode[] args) {
+        InvocationData data = ((SharedBytecodeParser) b).inlineInvocationData;
+        if (data == null) {
+            throw VMError.shouldNotReachHere("must not use NativeImageInlineDuringParsingPlugin when bytecode parser does not have InvocationData");
+        }
+
+        if (b.parsingIntrinsic()) {
+            /* We are not interfering with any intrinsic method handling. */
+            return null;
+        }
+        if (method.getAnnotation(NeverInline.class) != null || method.getAnnotation(NeverInlineTrivial.class) != null) {
+            return null;
+        }
+
+        if (method.getAnnotation(RestrictHeapAccess.class) != null || method.getAnnotation(Uninterruptible.class) != null ||
+                        b.getMethod().getAnnotation(RestrictHeapAccess.class) != null || b.getMethod().getAnnotation(Uninterruptible.class) != null) {
+            /*
+             * Caller or callee have an annotation that might prevent inlining. We don't check the
+             * exact condition but instead always bail out for simplicity.
+             */
+            return null;
+        }
+
+        CallSite callSite = new CallSite(toAnalysisMethod(b.getMethod()), b.bci());
+
+        InvocationResult inline;
+
+        if (analysis) {
+            InvocationResult existingResult = getResult(method, b.bci());
+            if (existingResult != null) {
```

Comment:
Why do we need this? I have a feeling that we should always do the analysis as it can give different results depending on the arguments to the function.